File: agario/bots/registry.py
"""Bot plugin registry and dynamic loader."""

from __future__ import annotations

import importlib
from collections.abc import Callable, Sequence

from .types import BotBrain, BotInitContext

BotFactory = Callable[[BotInitContext], BotBrain]


class BotRegistry:
    def __init__(self) -> None:
        self._factories: dict[str, BotFactory] = {}

    @property
    def names(self) -> tuple[str, ...]:
        return tuple(sorted(self._factories.keys()))

    def register(self, name: str, factory: BotFactory) -> None:
        key = (name or "").strip().lower()
        if not key:
            raise ValueError("Bot plugin name cannot be empty")
        if key in self._factories:
            raise ValueError(f"Duplicate bot plugin registration: {key}")
        self._factories[key] = factory

    def create(self, name: str, init_ctx: BotInitContext) -> BotBrain:
        key = name.strip().lower()
        factory = self._factories.get(key)
        if factory is None:
            available = ", ".join(self.names) or "<none>"
            raise ValueError(f"Unknown bot plugin '{key}'. Available: {available}")
        return factory(init_ctx)


def load_plugin_modules(module_names: Sequence[str], registry: BotRegistry) -> None:
    for module_name in module_names:
        name = module_name.strip()
        if not name:
            continue
        module = importlib.import_module(name)
        register_fn = getattr(module, "register", None)
        if not callable(register_fn):
            raise ValueError(f"Bot plugin module '{name}' must define register(registry)")
        register_fn(registry)

File: cpp_backend/src/bots/registry.cpp
#include "registry.hpp"

#include <algorithm>
#include <cctype>
#include <stdexcept>

namespace agario::bots {
namespace {

std::string normalize(const std::string& name) {
  std::string out;
  out.reserve(name.size());
  for (char ch : name) {
    if (std::isspace(static_cast<unsigned char>(ch))) {
      continue;
    }
    out.push_back(static_cast<char>(std::tolower(static_cast<unsigned char>(ch))));
  }
  return out;
}

}  // namespace

void BotRegistry::register_factory(const std::string& name, BotFactory factory) {
  std::string key = normalize(name);
  if (key.empty()) {
    throw std::invalid_argument("Bot plugin name cannot be empty");
  }
  if (factories_.find(key) != factories_.end()) {
    throw std::invalid_argument("Duplicate bot plugin registration: " + key);
  }
  factories_[key] = std::move(factory);
}

std::unique_ptr<BotBrain> BotRegistry::create(const std::string& name, const BotInitContext& init_ctx) const {
  std::string key = normalize(name);
  auto it = factories_.find(key);
  if (it == factories_.end()) {
    std::string available;
    for (const auto& entry : factories_) {
      if (!available.empty()) {
        available += ", ";
      }
      available += entry.first;
    }
    if (available.empty()) {
      available = "<none>";
    }
    throw std::invalid_argument("Unknown bot plugin '" + key + "'. Available: " + available);
  }
  return it->second(init_ctx);
}

std::vector<std::string> BotRegistry::names() const {
  std::vector<std::string> out;
  out.reserve(factories_.size());
  for (const auto& entry : factories_) {
    out.push_back(entry.first);
  }
  std::sort(out.begin(), out.end());
  return out;
}

}  // namespace agario::bots
